### PR TITLE
fix: Prevent focus ring from getting cut off in Popup.Body

### PR DESCRIPTION
**Pull request by mannycarrera4**
## Summary

Fixes: https://github.com/Workday/canvas-kit/issues/1682

category

### Release Note

- `Popup.Body` now has `8px` of padding. This addresses an issue introduced in v7 where focus rings were being clipped within the Body due to its built-in overflow; this padding allows the focus ring on any focusable element contained within the Body to be fully visible. The default space between the edge of the Card and the contents of the Body is still `32px`.
- `Popup.Heading` now has `8px` of padding to keep it visually aligned with the updated `Popup.Body`. It also now has a `marginBottom` of `8px`, bringing the default vertical space between the Heading and the Body to `24px` (previously `16px`) in order to meet the component spec.
- The padding in `Popup.Card` has been reduced by `8px` from `32px` to `24px` to compensate for the padding added to `Popup.Body` and `Popup.Heading` and to preserve the component's existing spacing. If you've customized the padding for any `Popup`, `Modal`, or `Dialog` subcomponents, you may need to adjust your padding values to get the same results after upgrading.
- `Popup.Card`'s default element is now a `Flex` element (previously a `Stack`) and thus the `spacing` prop has been removed. It made more sense to use `Flex` given the presence of the Close Icon, which didn't make sense as a `Stack` item.
- Given the changes above, if you're creating a footer for your `Popup`, `Modal`, or `Dialog` (using an `HStack`, for instance) you'll need to update it with the following props/styles to align it with the updated Heading and Body: `padding="xxs" (8px)` and `marginTop="xxs" (8px)`.

---

Screen Shot 2022-09-08 at 19 49 50

### CODE REVIEW

File: modules/react/modal/stories/examples/CustomFocus.tsx [hunk 1]
```diff
@@ -12,31 +12,31 @@ export const CustomFocus = () => {
...
       <Modal.Overlay>
         <Modal.Card>
           <Modal.CloseIcon aria-label="Close" />
-          <Modal.Heading>Delete Item</Modal.Heading>
-          <Modal.Body>
+          <Modal.Heading>Modal Heading</Modal.Heading>
+          <Modal.Body paddingBottom="zero">
             <Box as="p" marginTop={0} marginBottom="m">
               Enter name to confirm deletion
             </Box>
```
Comment (RayRedGoose) [inline modules/react/modal/stories/examples/CustomFocus.tsx:29]: I'm curious why did you change heading and button texts? New heading text is not connected to content it has. Maybe text should be abstract text as you had for heading and buttons, such it has in CustomTarget story.

Comment (mannycarrera4) [reply modules/react/modal/stories/examples/CustomFocus.tsx:29]: Yeah, i was just trying to go off of this https://chakra-ui.com/docs/components/modal/usage#usage which is very generic

Comment (mannycarrera4) [reply modules/react/modal/stories/examples/CustomFocus.tsx:29]: I can add some lorem ipsum!

Comment (RayRedGoose) [reply modules/react/modal/stories/examples/CustomFocus.tsx:29]: Also, it seems like @emroller16 did something close to this: https://github.com/Workday/canvas-kit/pull/1621/files

Comment (mannycarrera4) [reply modules/react/modal/stories/examples/CustomFocus.tsx:29]: Yeah, I followed up with her, she's on PTO, if she wants to commit that change, I'll go ahead and keep the stories as they where and let her make the change!

Comment (RayRedGoose) [reply modules/react/modal/stories/examples/CustomFocus.tsx:29]: Awesome! 👍

**Comment by josh-bagwell** [comment]
Manny and I discussed alternative approaches. One option was to move the buttons outside the `Modal.Body` component and that fixes the issue. However, it doesn't address the issue of links/buttons losing focus ring in `Modal.Body`. It looks as if the `overflow-y` is causing the focus ring issue.

**Comment by mannycarrera4** [comment]
### do not merged until `7.3` is released

File: modules/react/modal/lib/ModalBody.tsx [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+
+import {createSubcomponent, ExtractProps} from '@workday/canvas-kit-react/common';
+import {Card} from '@workday/canvas-kit-react/card';
+import {Popup, usePopupModel} from '@workday/canvas-kit-react/popup';
+
+export const ModalBody = createSubcomponent('div')({
+  displayName: 'Modal.Body',
+  modelHook: usePopupModel,
+})<ExtractProps<typeof Card.Body>>(elemProps => {
+  return <Popup.Body overflowY="auto" padding={'l'} paddingTop="zero" {...elemProps} />;
```
Comment (jamesfan) [inline modules/react/modal/lib/ModalBody.tsx:11]: I noticed the vertical spacing of a "default" Modal looks a bit off:

default-modal

Note the large vertical space between the Modal Body and the Modal Footer (`48px` vs `24px` in the spec). Even the vertical space between the Modal Heading and Modal Body is smaller than it should be (`16px` vs `24px` in the spec) -- this issue existed even before this PR.

I'm using the basic version of the Modal to render this example with almost no props applied to any of the subcomponents:

```
<Modal.Card>
  <Modal.CloseIcon aria-label="Close" />
  <Modal.Heading>Test</Modal.Heading>
  <Modal.Body>
    This is a test.
  </Modal.Body>
  <Modal.Footer>
    <Modal.CloseButton as={DeleteButton}>
      Delete
    </Modal.CloseButton>
    <Modal.CloseButton>Cancel</Modal.CloseButton>
  </Modal.Footer>
</Modal.Card>
```

What do you think about adjusting the default padding of the Modal subcomponents to match the `24px` of vertical spacing between the Header, Body, and Footer from the spec? I see there's a bottom padding on ModalBody Popup.Body, but that the top padding has been set to `zero`. Seems like we should add top padding to prevent content at the top of the Body from being cut off as well? We could add `paddingY="m"` to the ModalBody and remove all other vertical padding/margin from the subcomponents.

File: modules/react/modal/stories/stories_testing.tsx [hunk 3]
```diff
@@ -101,17 +101,17 @@ export const WithRadioButtons = () => {
...
-                name="contact"
-                data-testid="radiogroup"
-                value={value}
-                onChange={value => setValue(String(value))}
-              >
-                <Radio id="1" value="email" label="E-mail" />
-                <Radio id="2" value="phone" label="Phone" />
-              </RadioGroup>
-            </VStack>
+            {/* <VStack spacing="s"> */}
```
Comment (jamesfan) [inline modules/react/modal/stories/stories_testing.tsx:104]: This commented out `VStack` should be removed (or added back in).

File: modules/react/popup/lib/PopupCard.tsx [hunk 4]
```diff
@@ -64,7 +64,7 @@ export const PopupCard = createSubcomponent('div')({
   }, [model.state.placement]);
 
   // As is a Stack that will render an element of `Element`
-  const As = useConstant(() => Stack.as(Element));
+  const As = useConstant(() => Flex.as(Element));
```
Comment (jamesfan) [inline modules/react/popup/lib/PopupCard.tsx:67]: I noticed your comment about this in the PR summary:

> We've changed the base element on Popup.Card to use Flex instead of Stack to prevent breaking changes. If you were using spacing on that component, you'll have to use FlexStyleProps instead.

Can you clarify why this change was made? Which breaking changes are you referring to?

File: modules/react/modal/lib/ModalBody.tsx (hunk 2 above)
Comment (mannycarrera4) [reply modules/react/modal/lib/ModalBody.tsx:11]: That's perfect, thanks for that suggestion James!

File: modules/react/popup/lib/PopupCard.tsx (hunk 4 above)
Comment (jamesfan) [reply modules/react/popup/lib/PopupCard.tsx:67]: Just noticed this bit from the [7.4.0 Announcement](https://canvas.workday.com/whats-new/announcements#canvas-kit-740-release):

> As part of the Modal focus ring bug fix, each section (Heading, Body, and Footer) was separated to have their own padding and margins. Because of this change, the Stack in Modal.Card removed the margin top for the first element (Heading) in the Modal. As a result, the underlying Stack element for Modal.Card has also been replaced with the Flex element to add the 24px margin top back to the Modal Heading. Teams using the spacing prop on Modal.Card will need to remove it. If vertical alignment is required, teams should implement a VStack element.

Comment (mannycarrera4) [reply modules/react/popup/lib/PopupCard.tsx:67]: Yeah i would love some word smithing for this. What happened was because i moved margin to the heading element, `Stack` has a selector that was overriding this margin and setting it to `marginTop: 0` so I had to change this to use flex stack so that it would avoid this.

**Comment by jamesfan** [comment]
I chatted with @mannycarrera4 about this, and IMO it makes sense to apply these fixes to `Popup` instead of `Modal`. As it currently stands, this PR will fix the cut-off focus rings in `Modal`, but not in `Popup` or anything else which uses `Popup` (such as `Toast`). In order to address the original issue (#1682 which itself references the `Popup` and `Toast` components rather than `Modal`), we should fix the issue once at the source in `Popup` which will then propagate the fix to all other Popups (`Modal`, `Toast`, etc).

File: modules/react/popup/lib/PopupCard.tsx [hunk 5]
```diff
@@ -71,12 +70,11 @@ export const PopupCard = createSubcomponent('div')({
       as={As}
       transformOrigin={transformOrigin}
       position="relative"
-      padding="l"
       depth={5}
       maxWidth={`calc(100vw - ${space.l})`}
-      spacing={0}
```
Comment (mannycarrera4) [inline modules/react/popup/lib/PopupCard.tsx:77]: We can keep the underlying element a `Stack` however, if we keep this spacing="0" this will remove marginTop on the first element in the stack. I suggest we either remove this prop or change the underlying element to be `Flex`

File: modules/react/popup/lib/PopupCard.tsx [hunk 6]
```diff
@@ -71,12 +70,11 @@ export const PopupCard = createSubcomponent('div')({
...
       flexDirection="column"
       minHeight={0}
+      padding="zero"
```
Comment (mannycarrera4) [inline modules/react/popup/lib/PopupCard.tsx:77]: Move the padding from the `Card` to `Popup.Body` and margin to `Popup.Heading`

File: modules/react/popup/stories/examples/Basic.tsx [hunk 7]
```diff
@@ -27,11 +27,13 @@ export const Basic = () => {
     <Popup model={model}>
       <Popup.Target as={DeleteButton}>Delete Item</Popup.Target>
       <Popup.Popper placement="bottom">
-        <Popup.Card width={400} padding="s">
```
Comment (mannycarrera4) [inline modules/react/popup/stories/examples/Basic.tsx:30]: I removed this padding, to keep the same basic  styles at Modal

File: modules/react/modal/stories/examples/Basic.tsx [hunk 8]
```diff
@@ -17,7 +17,7 @@ export const Basic = () => {
           <Modal.CloseIcon aria-label="Close" />
           <Modal.Heading>Delete Item</Modal.Heading>
           <Modal.Body>
-            <Box as="p" marginTop={0} marginBottom="m">
+            <Box as="p" marginTop="zero" marginBottom="m">
```
Comment (mannycarrera4) [inline modules/react/modal/stories/examples/Basic.tsx:20]: consistency

File: modules/react/popup/lib/PopupBody.tsx [hunk 9]
```diff
@@ -8,5 +8,7 @@ export const PopupBody = createSubcomponent('div')({
   displayName: 'Popup.Body',
   modelHook: usePopupModel,
 })<ExtractProps<typeof Card.Body>>(elemProps => {
-  return <Card.Body overflowY="auto" {...elemProps} />;
+  return (
+    <Card.Body overflowY="auto" paddingX={'l'} paddingTop="s" paddingBottom="l" {...elemProps} />
```
Comment (mannycarrera4) [inline modules/react/popup/lib/PopupBody.tsx:12]: This is the main change, I removed the padding from `Popup.Card` and added it to the Body so that the focus ring is visible.

Comment (jamesfan) [reply modules/react/popup/lib/PopupBody.tsx:12]: Hmm, I'm seeing issues with our approach of adding vertical padding to the Body. Go to the Modal Body Overflow story and adjust the frame for the story to trigger overflow on the Body. Scroll down the Body and note how the content scrolls out of view at the top of the Body right below the Heading:

overflow-scrolling

File: modules/react/modal/stories/examples/BodyOverflow.tsx [hunk 10]
```diff
@@ -43,7 +42,7 @@ export const BodyOverflow = () => {
             <p>Are you sure you want to delete the item?</p>
             <p>Are you sure you want to delete the item?</p>
           </Modal.Body>
-          <HStack spacing="s" paddingTop="s">
+          <HStack spacing="s" paddingTop="s" paddingX="l" paddingBottom="l">
```
Comment (jamesfan) [inline modules/react/modal/stories/examples/BodyOverflow.tsx:45]: Now that we've removed `Modal.Footer`, it feels odd to me to require consumers to know they need `paddingX="l"` and `paddingBottom="l"` to get their `HStack` Footer to render with complementary padding to match `Modal.Body` and `Modal.Heading`. Previously, this was padding was baked into `Modal.Footer`.

Comment (mannycarrera4) [reply modules/react/modal/stories/examples/BodyOverflow.tsx:45]: if we make this a Minor release, it makes sense to add this back in. But I agree the consumer would have to know what padding to add

File: modules/react/modal/lib/ModalBody.tsx (hunk 2 above)
Comment (jamesfan) [reply modules/react/modal/lib/ModalBody.tsx]: We're going with Alan's approach instead (see top-level PR comment).

File: modules/react/popup/lib/PopupCard.tsx (hunk 5 above)
Comment (jamesfan) [reply modules/react/popup/lib/PopupCard.tsx:77]: We're changing PopupCard to use `Flex` instead of `Stack`. `Stack` doesn't work conceptually given the presence of the Popup.CloseIcon within the container (CloseIcon doesn't make sense as a `Stack` child).

File: modules/react/popup/lib/PopupBody.tsx (hunk 9 above)
Comment (jamesfan) [reply modules/react/popup/lib/PopupBody.tsx]: We've decided a small reduction in the vertical spacing between the Heading and Body as you scroll down the Body is acceptable (up to 8px reduction given the 8px of padding in Popup.Body, see Alan's approach in the top-level PR comment).

File: modules/react/modal/stories/examples/BodyOverflow.tsx (hunk 10 above)
Comment (jamesfan) [reply modules/react/modal/stories/examples/BodyOverflow.tsx]: For now, the consumer will need to add their own footer (using an `HStack`, for instance) with the appropriate paddings and margins to align with the Heading and Body subcomponents (see screenshot in top-level PR comment). We'll update Popup, Modal, and Dialog examples to include example footers with the needed padding and margin.

File: modules/react/modal/stories/examples/BodyOverflow.tsx [hunk 11]
```diff
@@ -18,8 +17,10 @@ export const BodyOverflow = () => {
           <Modal.CloseIcon aria-label="Close" />
           <Modal.Heading>Delete Item</Modal.Heading>
           <Modal.Body>
+            <PrimaryButton>FOCUSSSSSSSSSS</PrimaryButton>
```
Comment (jamesfan) [inline modules/react/modal/stories/examples/BodyOverflow.tsx:20]: Remove the test `PrimaryButton`s here (there are a few of them) and the `PrimaryButton` import at the top.

File: modules/react/popup/lib/PopupCard.tsx [hunk 12]
```diff
@@ -64,19 +65,18 @@ export const PopupCard = createSubcomponent('div')({
   }, [model.state.placement]);
 
   // As is a Stack that will render an element of `Element`
```
Comment (jamesfan) [inline modules/react/popup/lib/PopupCard.tsx:67]:
```suggestion
  // As is a Flex that will render an element of `Element`
```

**Review by jamesfan — APPROVED** [review]
